src/states.jl

```diff
@@ -0,0 +1,252 @@
+"""
+    get_vertical_states(model_type; snow = false, glacier = false)
+
+Function to extract all required vertical states, given a certain model type. Passes the snow
+and glacier options only for the `sbm` model_type. Returns a tuple with the required states
+(internal names as symbols)
+"""
+function get_vertical_states(model_type::String; snow = false, glacier = false)
+    if model_type == "sbm" || model_type == "sbm_gwf"
+        if snow && glacier
+            vertical_states = (
+                :satwaterdepth,
+                :snow,
+                :tsoil,
+                :ustorelayerdepth,
+                :snowwater,
+                :canopystorage,
+                :glacierstore,
+            )
+        elseif snow
+            vertical_states = (
+                :satwaterdepth,
+                :snow,
+                :tsoil,
+                :ustorelayerdepth,
+                :snowwater,
+                :canopystorage,
+            )
+        else
+            vertical_states = (:satwaterdepth, :ustorelayerdepth, :canopystorage)
+        end
+    elseif model_type == "hbv"
+        vertical_states = (
+            :soilmoisture,
+            :snow,
+            :snowwater,
+            :upperzonestorage,
+            :lowerzonestorage,
+            :interceptionstorage,
+        )
+    elseif model_type == "flextopo"
+        vertical_states = (
+            :snow,
+            :snowwater,
+            :interceptionstorage,
+            :hortonpondingstorage,
+            :hortonrunoffstorage,
+            :rootzonestorage,
+            :faststorage,
+            :slowstorage,
+        )
+    elseif model_type == "sediment"
+        vertical_states = ()
+    else
+        throw(ArgumentError("Unknown model_type provided (`$model_type`)"))
+    end
+    return vertical_states
+end
+
+"""
+    add_to_required_states(required_states, key_entry, state_list)
+
+Function to iterate through the several lists to add them to the `required_states` variables.
+This is a tuple of tuples, that contians the correct "prefix" to the state category (provided
+by the `key_entry` variable). The this is added in front of each element in the list of states
+(`state_list`), and the required_states tuple is returned.
+"""
+function add_to_required_states(
+    required_states::Tuple,
+    key_entry::Tuple,
+    state_list::Union{Nothing,Tuple},
+)
+
+    if !isnothing(state_list)
+        for state in state_list
+            required_states = (required_states..., (key_entry..., state))
+        end
+    end
+    return required_states
+end
```

Comment:
```suggestion
"""
    add_to_required_states(required_states::Tuple, key_entry::Tuple, states::Union{Nothing,Tuple})

Function to iterate through the list of `states` and add these to the `required_states`
variable. This is a tuple of tuples, that contains the correct "prefix" to the state
category (provided by the `key_entry` variable). This is added in front of each element in
the list of `states`, and the `required_states` tuple is returned.
"""
function add_to_required_states(
    required_states::Tuple,
    key_entry::Tuple,
    states::Tuple,
)
    for state in states
        required_states = (required_states..., (key_entry..., state))
    end

    return required_states
end

add_to_required_states(required_states::Tuple, key_entry::Tuple, states::Nothing,) = required_states
```

I suggest to make use of dispatch in this case.